playground/backend/internal/code_processing/code_processing.go

```diff
@@ -0,0 +1,218 @@
+// - In case of compile step is failed saves playground.Status_STATUS_COMPILE_ERROR as cache.Status and compile logs as cache.CompileOutput into cache.
+// - In case of compile step is completed with no errors saves compile output as cache.CompileOutput into cache.
+// - In case of run step is failed saves playground.Status_STATUS_RUN_ERROR as cache.Status and run logs as cache.RunError into cache.
+// - In case of run step is completed with no errors saves playground.Status_STATUS_FINISHED as cache.Status and run output as cache.RunOutput into cache.
+// At the end of this method deletes all created folders.
+func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, compileBuilder *executors.CompileBuilder, pipelineId uuid.UUID, env *environment.Environment, sdk pb.Sdk) {
+	ctxWithTimeout, cancelByTimeoutFunc := context.WithTimeout(ctx, env.ApplicationEnvs.PipelineExecuteTimeout())
+	defer func(lc *fs_tool.LifeCycle) {
+		cancelByTimeoutFunc()
+		cleanUp(pipelineId, lc)
+	}(lc)
+
+	errorChannel := make(chan error, 1)
+	dataChannel := make(chan interface{}, 1)
+	doneChannel := make(chan bool, 1)
+	cancelChan := make(chan bool, 1)
+
+	go cancelCheck(ctx, pipelineId, cancelChan, cacheService)
+
+	// build executor for validate and compile steps
+	exec := compileBuilder.Build()
+
+	// validate
+	logger.Infof("%s: Validate() ...\n", pipelineId)
+	validateFunc := exec.Validate()
+	go validateFunc(doneChannel, errorChannel)
+
+	if !processStep(ctxWithTimeout, pipelineId, cacheService, cancelChan, doneChannel, nil, errorChannel, pb.Status_STATUS_VALIDATION_ERROR, pb.Status_STATUS_COMPILING) {
+		return
+	}
+
+	// compile
+	logger.Infof("%s: Compile() ...\n", pipelineId)
+	compileCmd := exec.Compile(ctxWithTimeout)
+	go func(doneCh chan bool, errCh chan error, dataCh chan interface{}) {
+		data, err := compileCmd.CombinedOutput()
+		dataCh <- data
+		if err != nil {
+			errCh <- err
+			doneCh <- false
+		} else {
+			doneCh <- true
+		}
+	}(doneChannel, errorChannel, dataChannel)
+
+	if !processStep(ctxWithTimeout, pipelineId, cacheService, cancelChan, doneChannel, dataChannel, errorChannel, pb.Status_STATUS_COMPILE_ERROR, pb.Status_STATUS_EXECUTING) {
+		return
+	}
+
+	runBuilder, err := setuper.SetupRunBuilder(pipelineId, lc, sdk, env, compileBuilder)
+	if err != nil {
+		logger.Errorf("%s: error during setuper runBuilder: %s\n", pipelineId, err.Error())
+		cacheService.SetValue(ctxWithTimeout, pipelineId, cache.Status, pb.Status_STATUS_ERROR)
+		return
+	}
+
+	// build executor for run step
+	exec = runBuilder.Build()
+
+	// run
+	logger.Infof("%s: Run() ...\n", pipelineId)
+	runCmd := exec.Run(ctxWithTimeout)
+	go func(doneCh chan bool, errCh chan error, dataCh chan interface{}) {
+		data, err := runCmd.CombinedOutput()
+		dataCh <- data
+		if err != nil {
+			errCh <- err
+			doneCh <- false
+		} else {
+			doneCh <- true
+		}
+	}(doneChannel, errorChannel, dataChannel)
```

Comment:
I see this function is also used above. Can we define it separately and then call here and above?

Reply:
Added one more function to eliminate duplicates.